crates/runtime/src/datafusion/extension/cache_invalidation.rs

```diff
@@ -0,0 +1,537 @@
+        Vec::new()
+    }
+
+    fn fmt_for_explain(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "CacheInvalidationNode: table={}", self.table)
+    }
+
+    fn with_exprs_and_inputs(&self, exprs: Vec<Expr>, inputs: Vec<LogicalPlan>) -> Result<Self> {
+        assert_eq!(inputs.len(), 1, "should have one input");
+        assert_eq!(exprs.len(), 0, "should have no expressions");
+        let Some(input) = inputs.into_iter().next() else {
+            unreachable!("should have one input");
+        };
+        Ok(Self {
+            input,
+            table: self.table.clone(),
+            caching: Weak::clone(&self.caching),
+        })
+    }
+}
+
+impl PartialEq<CacheInvalidationNode> for CacheInvalidationNode {
+    fn eq(&self, other: &CacheInvalidationNode) -> bool {
+        self.input == other.input && self.table == other.table
+    }
+}
+
+impl Eq for CacheInvalidationNode {}
+
+impl Hash for CacheInvalidationNode {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.input.hash(state);
+        self.table.hash(state);
+    }
+}
+
+/// Physical [`ExecutionPlan`] that wraps a write operation and invalidates cache after successful completion.
+pub(crate) struct CacheInvalidationExec {
+    input_exec: Arc<dyn ExecutionPlan>,
+    table: TableReference,
+    caching: Weak<Caching>,
+}
+
+impl CacheInvalidationExec {
+    pub(crate) fn new(
+        input_exec: Arc<dyn ExecutionPlan>,
+        table: TableReference,
+        caching: Weak<Caching>,
+    ) -> Self {
+        Self {
+            input_exec,
+            table,
+            caching,
+        }
+    }
+
+    fn invalidate_cache_for_table(table: &TableReference, caching: &Weak<Caching>) {
+        if let Some(cache) = caching.upgrade() {
+            if let Err(e) = cache.invalidate_for_table(table.clone()) {
+                tracing::warn!("Failed to invalidate cache for table {table}: {e}");
+            } else {
+                tracing::trace!("Successfully invalidated cache for table {table}");
+            }
+        } else {
+            tracing::debug!(
+                "Cache reference for table {table} could not be upgraded; cache may have been dropped"
+            );
+        }
+    }
+}
+
+impl std::fmt::Debug for CacheInvalidationExec {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.debug_struct("CacheInvalidationExec")
+            .field("table", &self.table)
+            .finish_non_exhaustive()
+    }
+}
+
+impl DisplayAs for CacheInvalidationExec {
+    fn fmt_as(&self, t: DisplayFormatType, f: &mut fmt::Formatter) -> fmt::Result {
+        match t {
+            DisplayFormatType::Default | DisplayFormatType::Verbose => {
+                write!(f, "CacheInvalidationExec: table={}", self.table)
+            }
+            DisplayFormatType::TreeRender => {
+                write!(f, "CacheInvalidationExec")
+            }
+        }
+    }
+}
+
+impl ExecutionPlan for CacheInvalidationExec {
```

Comment:
Please double check that you've implemented all trat methods (where it makes sense to)

Reply:
Thank you, confirm I've reviewed and added missing methods (where is makes sense to)